lib/immutable-struct.rb

```diff
@@ -67,9 +85,37 @@ def self.new(*attributes,&block)
     end
     klass.class_exec(&block) unless block.nil?
     imethods = klass.instance_methods(include_super=false)
-    klass.class_exec(imethods) do |imethods|
+    klass.class_exec(imethods, attributes) do |imethods, attributes|
+
+      derived_methods = imethods - attributes - [:to_h]  #to_h is never a derived method, despite its 0-arity
+
+      ##
+      # :method: attributes_to_h
+      # Return a +Hash+ with only those attributes defined in the class constructor.
+      define_method(:attributes_to_h) do
+        attributes.each_with_object({}) do |attr, hash|
+          ivar = attr.to_s.prepend(?@).to_sym
+          hash[attr] = self.instance_variable_get(ivar)
+        end
+      end
+
+      ##
+      # :method: derived_to_h
+      # Return a +Hash+ of the results all zero-arity methods that are not attributes
+      # defined in the class constructor 
+      define_method(:derived_to_h) do
+        derived_methods.each_with_object({}) do |derived_method, hash|
+          hash[derived_method] = self.send(derived_method) if 0 == self.method(derived_method).arity
+        end
+      end
+
+      alias_method :__to_h__, :to_h if method_defined?(:to_h)
+      ##
+      # :method: to_h
+      # Return a +Hash+ with the result of merging +attributes_to_h+ and +derived_to_h+.
+      # This method can also be overriden so that, for example, only +attributes_to_h+ is used.
       define_method(:to_h) do
-        imethods.inject({}){ |hash, method| hash.merge(method.to_sym => self.send(method)) }
+        self.respond_to?(:__to_h__) ? __to_h__ : attributes_to_h.merge(derived_to_h)
```

Comment:
This `alias_method` handles the case where `#to_h` is overriden by the user:

```ruby
Klass = ImmutableStruct.new(:flappy) do
        def lawsuit
          'pending'
        end
        def to_h
          attributes_to_h  #or another custom logic
        end
      end
```

In the example above, the block is passed to the `ImmutableStruct` constructor and evaluated with `klass.class_exec(&block)` [just before](https://github.com/nandosola/immutable-struct/blob/747e744f9581e07d4265fcdc236dc4df1a1083e0/lib/immutable-struct.rb#L86) defining the `#attributes_to_h`, `#derived_to_h` and `#to_h` instance methods inside `klass.class_exec(imethods, attributes) do |imethods, attributes| ...`

So, if the user has decided to override `#to_h` with some custom conversion (or just calling `#attributes_to_h`) inside an `ImmutableStruct`, then it has to be aliased so that the original `#to_h` method is invoked, thus overriding the default `attributes_to_h.merge(derived_to_h)` .

In my application I'm invoking `#to_h` in different serializable domain classes other than `ImmutableStruct`, so IMHO this would make sense from a duck-typing point of view. TBH, I thought you meant this when we discussed issue #4 a few days ago:

> Then, if you don't wish to have the default implementation of to_h, you have a few options:
> * alias it to attributes_to_h
> * override it completely
> * use attribtues_to_h and derived_to_h however you like to pick and choose

So in my PR the two latter options are possible thanks to the user-defined `#to_h` being aliased to `#__to_h__`, while the first one won't work because `#to_h` cannot be aliased before it's defined, as I stated above.

However, I'm not making a case for anything here. So I'm cool with whatever you decide :-)